x/qgb/orchestrator/querier.go

```diff
@@ -0,0 +1,269 @@
+	qgbGRPC, err := grpc.Dial(qgbRpcAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	trpc, err := http.New(tendermintRpc, "/websocket")
+	if err != nil {
+		return nil, err
+	}
+	err = trpc.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return &querier{
+		qgbRPC:        qgbGRPC,
+		logger:        logger,
+		tendermintRPC: trpc,
+	}, nil
+}
+
+func (q *querier) QueryDataCommitments(
+	ctx context.Context,
+	commit string,
+) ([]types.MsgDataCommitmentConfirm, error) {
+	queryClient := types.NewQueryClient(q.qgbRPC)
+
+	confirmsResp, err := queryClient.DataCommitmentConfirmsByCommitment(
+		ctx,
+		&types.QueryDataCommitmentConfirmsByCommitmentRequest{
+			Commitment: commit,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return confirmsResp.Confirms, nil
+}
+
+func (q *querier) QueryTwoThirdsDataCommitmentConfirms(
+	ctx context.Context,
+	timeout time.Duration,
+	commitment string,
+) ([]types.MsgDataCommitmentConfirm, error) {
+	// query for the latest valset (sorted for us already)
+	queryClient := types.NewQueryClient(q.qgbRPC)
+	lastValsetResp, err := queryClient.LastValsetRequests(ctx, &types.QueryLastValsetRequestsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(lastValsetResp.Valsets) < 1 {
+		return nil, errors.New("no validator sets found")
+	}
+
+	valset := lastValsetResp.Valsets[0]
+
+	// create a map to easily search for power
+	vals := make(map[string]types.BridgeValidator)
+	for _, val := range valset.Members {
+		vals[val.GetEthereumAddress()] = val
+	}
+
+	majThreshHold := valset.TwoThirdsThreshold()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil
+		case <-time.After(timeout):
+			return nil, fmt.Errorf("failure to query for majority validator set confirms: timout %s", timeout)
+		default:
+			currThreshHold := uint64(0)
+			confirmsResp, err := queryClient.DataCommitmentConfirmsByCommitment(
+				ctx,
+				&types.QueryDataCommitmentConfirmsByCommitmentRequest{
+					Commitment: commitment,
+				},
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, dataCommitmentConfirm := range confirmsResp.Confirms {
+				val, has := vals[dataCommitmentConfirm.EthAddress]
+				if !has {
+					return nil, fmt.Errorf(
+						"dataCommitmentConfirm signer not found in stored validator set: address %s nonce %d",
+						val.EthereumAddress,
+						valset.Nonce,
+					)
+				}
+				currThreshHold += val.Power
+			}
+
+			if currThreshHold >= majThreshHold {
+				return confirmsResp.Confirms, nil
+			}
+			q.logger.Debug(
+				"foundDataCommitmentConfirms",
+				fmt.Sprintf(
+					"total power %d number of confirms %d",
+					currThreshHold,
+					len(confirmsResp.Confirms),
+				),
+			)
+		}
+		// TODO: make the timeout configurable
+		time.Sleep(time.Second * 30)
+	}
+}
+
+func (q *querier) QueryTwoThirdsValsetConfirms(
+	ctx context.Context,
+	timeout time.Duration,
+	valset types.Valset,
```

Comment:
I also think that we can simply rely on the context as long as we document that it will run forever until configured to stop.

Reply:
yes, sure. I guess this can be handled when we have the worker pool design.

Reply:
And also would allow us to optimize a lot about the workflows we're calling for different queries